drivers/timer/cortex_m_systick.c

```diff
@@ -32,19 +45,40 @@ static u32_t cycle_count;
 
 static u32_t announced_cycles;
 
-static volatile u32_t ctrl_cache; /* overflow bit clears on read! */
+static volatile u32_t overflow_cyc;
 
 static u32_t elapsed(void)
 {
-	u32_t val, ov;
+	u32_t val, ctrl1, ctrl2;
+
+	/* SysTick is infuriatingly racy.  The counter wraps at zero
+	 * automatically, setting a 1 in the overflow bit of the CTRL
+	 * register when it does.  But reading the control register
+	 * automatically resets that bit, so we need to save it for
+	 * future calls.  And ordering is critical and race-prone: if
+	 * we read CTRL first, then it is possible for VAL to wrap
+	 * after that read but before we read VAL and we'll miss the
+	 * overflow.  If we read VAL first, then it can wrap after we
+	 * read it and we'll see an "extra" overflow in CTRL.  And we
+	 * want to handle multiple overflows, so we effectively must
+	 * read CTRL first otherwise there will be no way to detect
+	 * the double-overflow if called at the end of a cycle.  There
+	 * is no safe algorithm here, so we split the difference by
+	 * reading CTRL twice, suppressing the second overflow bit if
+	 * VAL was "about to overflow".
+	 */
+	ctrl1 = SysTick->CTRL;
+	val = SysTick->VAL & COUNTER_MAX;
+	ctrl2 = SysTick->CTRL;
 
-	do {
-		val = SysTick->VAL & COUNTER_MAX;
-		ctrl_cache |= SysTick->CTRL;
-	} while (SysTick->VAL > val);
+	overflow_cyc += (ctrl1 & SysTick_CTRL_COUNTFLAG_Msk) ? last_load : 0;
```

Comment:
What about reading ctrl and val 2 times (CTRL1, VAL1, CTRL2, VAL2)?
We will be real;y unlucky if we get 2 wraps during such read. So instead of assuming that the wrap bit in CTRL2 is valid depending on value of VAL, we can confirm that comparing values of VAL1 and VAL2.

Reply:
That would work too.  Really the code is mostly the same either way: check VAL and test against previous VAL vs. check VAL and test against a fixed threshold.  Both can race with essentially the same frequency (i.e. you'd have to stall the timer interrupt for a full LOAD cycle), which is low.  I picked the version that does three register reads instead of four out of general paranoia I guess (qemu is known to be glitchy on some platforms if you spam the timing registers).